File: C#-OOP/02.EncapsulationExercise/FootballTeamGenerator/Team.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace FootballTeamGenerator
{
    public class Team
    {
        private string name;
        private List<Player> players;

        public Team(string name)
        {
            Name = name;

            players = new List<Player>();
        }

        public string Name
        {
            get => name;

            private set
            {
                Validator.IsNullOrWhiteSpace(value);

                name = value;
            }
        }

        public double Rating => Math.Round(players.Average(p => p.Overall));

        public void AddPlayer(Player player)
        {
            players.Add(player);
        }

        public void RemovePlayer(string playerName)
        {
            if (players.Any(p => p.Name == playerName))
            {
                Player player = players.First(p => p.Name == playerName);

                players.Remove(player);
            }
            else
            {
                throw new InvalidOperationException($"Player {playerName} is not in {Name} team.");
            }
        }

        public override string ToString()
        {
            if (players.Count != 0)
            {
                return $"{Name} - {Rating}";
            }
            else
            {
                return $"{Name} - 0";
            }
        }
    }
}

File: Python-OOP/04Encapsulation-Exercise/03Football-Team-Generator/project/team.py
class Team:
    def __init__(self, name, rating):
        self.__name = name
        self.__rating = rating
        self.__players = []

    def add_player(self, new_player):
        if new_player in self.__players:
            return f'Player {new_player.name} has already joined'

        self.__players.append(new_player)
        return f'Player {new_player.name} joined team {self.__name}'

    def remove_player(self, player_name):
        for player in self.__players:
            if player.name == player_name:
                self.__players.remove(player)
                return player

        return f'Player {player_name} not found'

